# Exploración

Comenzamos cargando las librerías


In [1]:
import sys, platform
import pandas as pd
import numpy as np

print("Python", sys.version)
print("Platform", platform.platform())
print("pandas:", pd.__version__)
print("numpy:", np.__version__)

Python 3.12.10 (tags/v3.12.10:0cc8128, Apr  8 2025, 12:21:36) [MSC v.1943 64 bit (AMD64)]
Platform Windows-11-10.0.26100-SP0
pandas: 2.3.2
numpy: 2.3.3


Cargamos los datos

In [4]:
data = "C:/Users/hered/Desktop/TFM/TFM/data/PMC-Patients.csv"

df = pd.read_csv(data, dtype=str, encoding="utf-8", na_filter=False)

print("Filas:", len(df))
print("Columnas:", list(df.columns))

#visualizamos

try:
    display(df.head(3))
except NameError:
    print(df.head(3))

Filas: 167034
Columnas: ['patient_id', 'patient_uid', 'PMID', 'file_path', 'title', 'patient', 'age', 'gender', 'relevant_articles', 'similar_patients']


patient_id patient_uid      PMID                         file_path  \
0          0   7665777-1  33492400  comm/PMC007xxxxxx/PMC7665777.xml   
1          1   7665777-2  33492400  comm/PMC007xxxxxx/PMC7665777.xml   
2          2   7665777-3  33492400  comm/PMC007xxxxxx/PMC7665777.xml   

                                               title  \
0  Early Physical Therapist Interventions for Pat...   
1  Early Physical Therapist Interventions for Pat...   
2  Early Physical Therapist Interventions for Pat...   

                                             patient               age gender  \
0  This 60-year-old male was hospitalized due to ...  [[60.0, 'year']]      M   
1  A 39-year-old man was hospitalized due to an i...  [[39.0, 'year']]      M   
2  One week after a positive COVID-19 result this...  [[57.0, 'year']]      M   

                                   relevant_articles  \
0  {'32320506': 1, '32293716': 1, '23219649': 1, ...   
1  {'32320506': 1, '32293716': 1, '23219649': 1, ...   
2  {'32320506': 1, '32293716': 1, '23219649': 1, ...   

                                    similar_patients  
0  {'7665777-2': 2, '7665777-3': 2, '7665777-4': ...  
1  {'7665777-1': 2, '7665777-3': 2, '7665777-4': ...  
2  {'7665777-1': 2, '7665777-2': 2, '7665777-4': ...

Como exploración inicial, comenzamos viendo la distribución de gender y longitudes de la columna de patient

In [5]:
if "gender" in df.columns:
    print("Value conts de 'gender'")
    print(df["gender"].value_counts().head(10))
else:
    print("no existe")

Value conts de 'gender'
gender
M    87659
F    79375
Name: count, dtype: int64


In [6]:
if "patient" in df.columns:
    text_len = df["patient"].str.len()
    print("\nLongitud 'patient' - min/median/mean/max")
    print(int(text_len.min()), int(text_len.median()), round(text_len.mean(),1), int(text_len.max()))
else:
    print("no existe patient")


Longitud 'patient' - min/median/mean/max
55 2491 2765.3 95400


In [12]:
print("\nEjemplos de 'patient'")
for s in df.loc[df["patient"].str.len()>0, "patient"].head(2):
    print("-"*80)
    print(s.replace("\n"," "))


Ejemplos de 'patient'
--------------------------------------------------------------------------------
This 60-year-old male was hospitalized due to moderate ARDS from COVID-19 with symptoms of fever, dry cough, and dyspnea. We encountered several difficulties during physical therapy on the acute ward. First, any change of position or deep breathing triggered coughing attacks that induced oxygen desaturation and dyspnea. To avoid rapid deterioration and respiratory failure, we instructed and performed position changes very slowly and step-by-step. In this way, a position change to the 135° prone position () took around 30 minutes. This approach was well tolerated and increased oxygen saturation, for example, on day 5 with 6 L/min of oxygen from 93% to 97%. Second, we had to adapt the breathing exercises to avoid prolonged coughing and oxygen desaturation. Accordingly, we instructed the patient to stop every deep breath before the need to cough and to hold inspiration for better air di

#### Ejemplo del primer "patient" de nuestro dataset:

This 60-year-old male was hospitalized due to moderate ARDS from COVID-19 with symptoms of fever, dry cough, and dyspnea. We encountered several difficulties during physical therapy on the acute ward. First, any change of position or deep breathing triggered coughing attacks that induced oxygen desaturation and dyspnea. To avoid rapid deterioration and respiratory failure, we instructed and performed position changes very slowly and step-by-step. In this way, a position change to the 135° prone position () took around 30 minutes. This approach was well tolerated and increased oxygen saturation, for example, on day 5 with 6 L/min of oxygen from 93% to 97%. Second, we had to adapt the breathing exercises to avoid prolonged coughing and oxygen desaturation. Accordingly, we instructed the patient to stop every deep breath before the need to cough and to hold inspiration for better air distribution. In this manner, the patient performed the breathing exercises well and managed to increase his oxygen saturation. Third, the patient had difficulty maintaining sufficient oxygen saturation during physical activity. However, with close monitoring and frequent breaks, he managed to perform strength and walking exercises at a low level without any significant deoxygenation. Exercise progression was low on days 1 to 5, but then increased daily until hospital discharge to a rehabilitation clinic on day 10.

#### Segundo ejemplo:

A 39-year-old man was hospitalized due to an increasingly reduced general health condition, after persistent fever and dry cough for 2 weeks. The patient initially needed 4 L/min of oxygen, had a rapid and shallow breathing pattern at rest and became severely breathless during minor physical activities. In the beginning, physical therapy focused on patient education about dyspnea-relieving positions, the importance of regular mobilization, and deep-breathing exercises. However, it quickly became evident that his anxiety from fear of dying and worries about his future aggravated his dyspnea and vice versa. The patient was so dyspneic, anxious, and weak that he was barely able to walk to the toilet. To counter this vicious circle, the physical therapist actively listened to the patient, explained why he was experiencing breathlessness, and tested suitable positions to relieve his dyspnea. He seemed to benefit from the education and the relaxing breathing exercises, as seen on day 2, when his respiratory rate could be reduced from 30 breaths/min to 22 breaths/min and his oxygen saturation increased from 92% to 96% on 4 L/min oxygen after guiding him through some deep-breathing exercises. Over the next days, his dyspnea and anxiety started to alleviate and he regained his self-confidence. Therapy was progressively shifted to walking and strength training and the patient rapidly advanced to walk 350 m without a walking aid or supplemental oxygen before his discharge home.



Como hemos podido observar "age" viene como [[60.0, 'year']] y "gender" como M/F.  

El siguiente paso será normalizar "age" y "gender":

- age_col_norm (int) 
     
    "[[60.0, 'year']]"  -> 60
    "60" / "60.0"       -> 60
    Devuelve int o None si no encuentra un valor razonable (0<edad<130).

- sex_col_norm (male/female)

    Mapea variantes a {'male','female'}.
    Acepta: 'M','F','male','female','man','woman','hombre','mujer', etc.

para poder comparar con lo que se extraiga del texto.

In [13]:
import re

def parse_age_col(val:str):
    if not val or not isinstance(val, str):
        return None
    m = re.search(r'([0-9]+(?:\.[0-9]+)?)',val)
    if not m:
        return None
    try:
        num = float(m.group(1))
        if 0 < num < 130:
            return int(round(num))
    except:
        return None
    return None

def normalize_gender_col(val:str):
    if not val:
        return None
    v = val.strip().lower()
    if v in {"m", "male", "man", "hombre"}:
        return "male"
    if v in {"f", "female", "woman", "mujer"}:
        return "female"
    return None

#Creamos las columnas normalizadas:

df["age_col_norm"] = df.get("age", "").apply(parse_age_col)
df["sex_col_norm"] = df.get("gender", "").apply(normalize_gender_col)

#Vista
try:
    display(df[["age", "age_col_norm", "gender", "sex_col_norm"]].head(5))
except NameError:
    print(df[["age","age_col_norm","gender","sex_col_norm"]].head(5))

age  age_col_norm gender sex_col_norm
0  [[60.0, 'year']]          60.0      M         male
1  [[39.0, 'year']]          39.0      M         male
2  [[57.0, 'year']]          57.0      M         male
3  [[69.0, 'year']]          69.0      M         male
4  [[57.0, 'year']]          57.0      M         male

El siguiente paso será extraer edad y sexo desde la columna "patient" y vamos a compararlas con las columnas normalizadas

In [16]:
# %% [markdown]
# 📊 CELDA 7 — ANTES
# Objetivo: Aplicar la extracción a TODAS las filas y comparar con age_col_norm / sex_col_norm.

# %%
import pandas as pd
import numpy as np
import time

t0 = time.time()

rows = []
# itertuples es más rápido que iterrows
for r in df.itertuples(index=False):
    note_text      = getattr(r, "patient", "")
    patient_id     = getattr(r, "patient_id", "")
    age_col_norm   = getattr(r, "age_col_norm", None)
    sex_col_norm   = getattr(r, "sex_col_norm", None)

    ext = extract_age_sex_from_text(note_text)
    rows.append({
        "patient_id": patient_id,
        "age_col_norm": age_col_norm,
        "sex_col_norm": sex_col_norm,
        "age_from_text": ext.get("age_from_text"),
        "sex_from_text": ext.get("sex_from_text"),
        "sex_from_text_norm": ext.get("sex_from_text_norm"),
    })

cmp_df = pd.DataFrame(rows)

# Aseguramos tipos numéricos enteros comparables (nullable Int64 para manejar NaN)
cmp_df["age_col_norm_int"]   = pd.to_numeric(cmp_df["age_col_norm"], errors="coerce").astype("Int64")
cmp_df["age_from_text_int"]  = pd.to_numeric(cmp_df["age_from_text"], errors="coerce").astype("Int64")

# Coincidencias (True/False/NaN si falta algún lado)
cmp_df["age_match"] = np.where(
    cmp_df["age_from_text_int"].notna() & cmp_df["age_col_norm_int"].notna(),
    cmp_df["age_from_text_int"] == cmp_df["age_col_norm_int"],
    np.nan
)
cmp_df["sex_match"] = np.where(
    cmp_df["sex_from_text_norm"].notna() & cmp_df["sex_col_norm"].notna(),
    cmp_df["sex_from_text_norm"] == cmp_df["sex_col_norm"],
    np.nan
)

t1 = time.time()

print(f"Filas procesadas: {len(cmp_df)}")
print(f"Tiempo (s): {t1 - t0:.1f}")
print("age_match -> proporción de True (ignorando NaN):", cmp_df["age_match"].mean(skipna=True))
print("sex_match -> proporción de True (ignorando NaN):", cmp_df["sex_match"].mean(skipna=True))

# Muestra algunas filas para inspección
try:
    display(cmp_df.head(10))
except NameError:
    print(cmp_df.head(10))

# %% [markdown]
# ✅ CELDA 7 — DESPUÉS
# Deberías ver:
# - Filas procesadas ≈ 167,034
# - Proporciones de coincidencia para edad/sexo (ignorando NaNs)
# - Un vistazo a las primeras filas del DataFrame comparativo.


Filas procesadas: 167034
Tiempo (s): 18.5
age_match -> proporción de True (ignorando NaN): 0.9910785683988012
sex_match -> proporción de True (ignorando NaN): 0.4444444444444444


patient_id  age_col_norm sex_col_norm  age_from_text sex_from_text  \
0          0          60.0         male           60.0          None   
1          1          39.0         male           39.0          None   
2          2          57.0         male           57.0          None   
3          3          69.0         male           69.0          None   
4          4          57.0         male           57.0          None   
5          5          52.0         male           52.0          None   
6          6          59.0       female           59.0          None   
7          7          33.0       female           33.0          None   
8          8          66.0         male           66.0          None   
9          9          66.0         male           66.0          None   

  sex_from_text_norm  age_col_norm_int  age_from_text_int age_match  sex_match  
0               None                60                 60      True        NaN  
1               None                39                 39      True        NaN  
2               None                57                 57      True        NaN  
3               None                69                 69      True        NaN  
4               None                57                 57      True        NaN  
5               None                52                 52      True        NaN  
6               None                59                 59      True        NaN  
7               None                33                 33      True        NaN  
8               None                66                 66      True        NaN  
9               None                66                 66      True        NaN

kcjdcjn